### ISSUE TITLE
libcudnn_adv_train.so not found when using LSTM, unless I run a BatchNorm before

### ISSUE THREAD

**Issue by lemairecarl** [tensorflow/tensorflow#51521]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS Linux 7.9.2009
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.5.0
- Python version: 3.8
- Bazel version (if compiling from source): 3.7.2
- GCC/Compiler version (if compiling from source): GCC 9.3.0
- CUDA/cuDNN version: CUDA 11.1, cuDNN 8.2
- GPU model and memory: NVidia V100 16GB

**Describe the current behavior**

Please note that I have CUDA and cuDNN installed in custom directories. This causes problems in dynamically loading cuDNN libs. (To be able to use TF with custom CUDA/cuDNN paths, I had to patch the built wheel, by setting RPATHs on the `*.so` inside the wheel)

When running an LSTM (see repro code below), I get this error:
```
Could not load library libcudnn_adv_train.so.8. Error: libcudnn_ops_train.so.8: cannot open shared object file: No such file or directory
Please make sure libcudnn_adv_train.so.8 is in your library path!
Aborted (core dumped)
```
However, if I execute a BatchNorm layer right before, there is no error and I can use the LSTM. This is because, using BatchNorm triggers the load of `libcudnn_ops_train.so.8`, which works this way.

**Describe the expected behavior**

LSTM should work.

Maybe `libcudnn_ops_train.so.8` should be manually loaded by TF ?

I also expect not having the patch the wheel. If I set CUDA/cuDNN paths when configuring the build, why won't TF use these paths?

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```python
import tensorflow as tf

# If I uncomment this, there is no crash.
# bn = tf.keras.layers.BatchNormalization()
# inputs = tf.random.normal([32, 10, 8, 8])
# output = bn(inputs)

inputs = tf.random.normal([32, 10, 8])
lstm = tf.keras.layers.LSTM(4)
output = lstm(inputs)
```

**Comment by sushreebarsa** [comment]
@lemairecarl I tried to run your code on colab using TF v2.5, 2.6.0(latest stable version of TF ) and didn't face any error reported .Please find the gist[ here ](https://colab.research.google.com/gist/sushreebarsa/4b0fbb80dbc09306ad2d351c2a4c748c/untitled388.ipynb)for reference. We see that you are using GCC/Compiler version (if compiling from source): GCC 9.3.0 ,could you please try to use GCC 7.3.1 and let us know if the issue still persists ? Thank you!

**Comment by lemairecarl** [comment]
I think a very important aspect to be able to reproduce this issue, is to have installed cuDNN in a custom path. If it's installed in the default path, you will not be able to reproduce this issue.

I have access to GCC 7.3.0, do you think it will work?

**Comment by lemairecarl** [comment]
During the configure step of the build process, I give the custom paths where CUDA/cuDNN are installed on my system. However, the built wheel cannot be used with the custom cuDNN installation (unless the wheel is patched). Why doesn't the build process automatically add the corresponding RUNPATHs to the binaries in the wheel? (Such as `_pywrap_tensorflow_internal.so`)

**Comment by jvishnuvardhan** [comment]
@mihaimaruseac Can you please take a look at it? Thanks!

**Comment by lemairecarl** [comment]
I have used the following workaround: I patched the RPATHs of cuDNN binaries to include $ORIGIN.

Also, please note that on the system I use, RPATHs should be used instead of RUNPATHs. (The tensorflow build process uses RUNPATHs).

**Comment by mohantym** [comment]
Hi @lemairecarl !
Could you set the path for Cuda and cudnn file after installing TF 2.9 and Cuda 11.2 ,CudNN 8.1 in a fresh environment  and let us know.
```
export PATH="/usr/local/cuda-11.2/bin:$PATH"
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/cuda-11.2/lib64
```
[Reference](https://github.com/tensorflow/tensorflow/issues/41041#issuecomment-895059400)
